cmake/package_tooling.cmake

```diff
@@ -0,0 +1,128 @@
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
+#          set(OVERLAY_CONFIG foo.overlay)
+#          find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+#       the 'foo.overlay' must be specified using '-DOVERLAY_CONFIG=foo.overlay'
+
+cmake_minimum_required(VERSION 3.20)
+
+# add_custom_target and set_target_properties are not supported in script mode.
+# However, several Zephyr CMake modules create custom target for user convenience
+# like menuconfig, boards, shields, etc.
+# As we are not generating a build system with this tool, only running part of
+# the modules, then we simply override those functions to allow running those
+# modules.
+function(add_custom_target)
+  # This silence the error: 'add_custom_target command is not scriptable'
+endfunction()
+
+function(set_target_properties)
+  # This silence the error: 'set_target_properties command is not scriptable'
+endfunction()
+
+# Find last `-B` and `-S` instances.
+foreach(i RANGE ${CMAKE_ARGC})
+  if(CMAKE_ARGV${i} MATCHES "^-B(.*)")
+    set(argB ${CMAKE_MATCH_1})
+    set(argB_index ${i})
+  elseif()
+  elseif(CMAKE_ARGV${i} MATCHES "^-S(.*)")
+    set(argS_index ${i})
+    set(argS ${CMAKE_MATCH_1})
+  endif()
+endforeach()
+
+if(DEFINED argB_index)
+  if(DEFINED argB)
+    set(CMAKE_BINARY_DIR ${argB})
+  else()
+    # value of -B follows in next index
+    math(EXPR argB_value_index "${argB_index} + 1")
+    set(CMAKE_BINARY_DIR ${CMAKE_ARGV${argB_value_index}})
+  endif()
+endif()
+
+if(DEFINED argS_index)
+  if(DEFINED argS)
+    set(APPLICATION_SOURCE_DIR ${argS})
+  else()
+    # value of -S follows in next index
+    math(EXPR argS_value_index "${argS_index} + 1")
+    set(APPLICATION_SOURCE_DIR ${CMAKE_ARGV${argS_value_index}})
+  endif()
+endif()
+
+# List of base modules needed in order to ensure same behavior as the default
+# Zephyr build system during configure time.
+set(BASE_MODULES python west root zephyr_module boards shields arch configuration_files)
+
+# dts is not directly a dependency of Kconfig, but in Zephyr build system we
+# import devicetree properties into the Kconfig tree and therefore we must ensure
+# dts module is loaded in Kconfig mode.
+if(kconfig IN_LIST MODULES)
+  list(APPEND BASE_MODULES dts)
+endif()
+
+if(NOT DEFINED APPLICATION_SOURCE_DIR)
+  message(FATAL_ERROR
+    "Source directory not defined, please use '-S <path-to-source>' to the "
+    "application for which this tool shall run.\n"
+    "For example: -S ${ZEPHYR_BASE}/samples/hello_world"
+  )
+endif()
+if(DEFINED APPLICATION_SOURCE_DIR)
+  cmake_path(ABSOLUTE_PATH APPLICATION_SOURCE_DIR NORMALIZE)
+endif()
+
+cmake_path(ABSOLUTE_PATH CMAKE_BINARY_DIR NORMALIZE)
+set(CMAKE_CURRENT_BINARY_DIR ${CMAKE_BINARY_DIR})
+
+if(NOT DEFINED MODULES)
+  message(WARNING
+    "No MODULES defined, please invoke tooling with minimum one module"
+    " to execute in script mode."
```

Comment:
Could we list the scripts that are supported to run in script mode here?

Or are all scripts supported?

I'm concerned users will generate support questions like "can I run script X" in script mode?

Reply:
the intention with this script is to be able to run the Zephyr build system as is done in configure time (not generator time), and stop at any point before the `kernel.cmake` module.

That will then produce the configure time output from that module, as if it had been executed as part of Zephyr CMake `configure` time. This allows to examine the output.

Such output could be `.config`, `zephyr.dts`, `zephyr_modules.txt`, etc.

Especially useful for twister, which allows to do `dts` or `kconfig` and then filter the output to see if a given sample + board meets a pre-condition for running a full build.

Reply:
So skipping scripts that are not needed for the desired target is out-of-scope?

Reply:
> So skipping scripts that are not needed for the desired target is out-of-scope?

not sure I understand the question.
Can you elaborate.

Did you read the purpose of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/41302#issue-1083249333
> The Zephyr CMake package tooling script allows to invoke individual
Zephyr package modules (and their dependencies) in script mode without
running the complete CMake configure and build system generation stages.
>
> This means that for example devicetree module can be invoked for a given
sample and board to generate zephyr.dts identical to what is created
by the build system during a build.

and:
> This will allow other tooling, such as twister, to executed only a
subset of the build system like dts in order to do filtering before
determine if a complete build should be invoked.

Reply:
There are two strategies, either

1. run the exact same scripts in the exact same order and then stop when we have achieved the desired result.

2. run the desired target and rely on it including it's dependencies.

The PR description is describing 2. But this sentence is describing 2:

"the intention with this script is to be able to run the Zephyr build system as is done in configure time (not generator time), and stop at any point before the kernel.cmake module."

I'm wondering if the minimal execution that we get with 2. is in-scope, or if it is the simpler strategy 1. we are going for.

Reply:
ok, maybe `dependencies` in the PR description is misleading.

It's the execution order / dependencies seen from the regular Zephyr CMake configure time, not the internal Zephyr CMake module dependencies.

So the correct answer is 1.

Reply:
I have updated the description, and the code to make this more clear.

Reply:
This is now resolved. Execution can stop after any script in zephyr_default.cmake is executed.